rust-runtime/aws-smithy-client/src/timeout.rs

```diff
@@ -51,8 +59,257 @@ impl Settings {
+/// A struct containing everything needed to create a new [`TimeoutService`]
+pub struct TimeoutServiceParams {
+    /// The duration of timeouts created from these params
+    duration: Duration,
+    /// The kind of timeouts created from these params
+    kind: &'static str,
+    /// The AsyncSleep impl that will be used to create time-limited futures
+    async_sleep: Arc<dyn AsyncSleep>,
+}
+
+#[derive(Clone, Debug, Default)]
+/// A struct of structs containing everything needed to create new [`TimeoutService`]s
+pub struct AllTimeoutServiceParams {
+    /// Params used to create a new API call [`TimeoutService`]
+    pub api_call: Option<TimeoutServiceParams>,
+    /// Params used to create a new API call attempt [`TimeoutService`]
+    pub api_call_attempt: Option<TimeoutServiceParams>,
+}
+
+/// Convert a [`TimeoutConfig`] into an [`AllTimeoutServiceParams`] in order to create the set of
+/// [`TimeoutService`]s needed by a [`crate::Client`]
+pub fn generate_timeout_service_params_from_timeout_config(
+    timeout_config: &TimeoutConfig,
+) -> AllTimeoutServiceParams {
+    if let Some(async_sleep) = default_async_sleep() {
+        AllTimeoutServiceParams {
+            api_call: timeout_config
+                .api_call_timeout()
+                .map(|duration| TimeoutServiceParams {
+                    duration,
+                    kind: "API call (multiple attempts)",
+                    async_sleep: async_sleep.clone(),
+                }),
+            api_call_attempt: timeout_config.api_call_attempt_timeout().map(|duration| {
+                TimeoutServiceParams {
+                    duration,
+                    kind: "API call (single attempt)",
+                    async_sleep: async_sleep.clone(),
+                }
+            }),
+        }
+    } else {
+        let list_of_set_timeouts = timeout_config.list_of_set_timeouts();
+        let list_of_set_timeouts = list_of_set_timeouts.join(", ");
+
+        tracing::warn!(
+            "One or more timeouts were set ({}) but no default_async_sleep fn exists. \
+            Make sure the 'tokio-rt' feature is enabled if you want to set timeouts.",
+            list_of_set_timeouts
+        );
+
+        Default::default()
+    }
+}
+
+/// A service that wraps another service, adding the ability to set a timeout for requests
+/// handled by the inner service.
+#[derive(Clone, Debug)]
+pub struct TimeoutService<InnerService> {
+    inner: InnerService,
+    params: Option<TimeoutServiceParams>,
+}
+
+impl<InnerService> TimeoutService<InnerService> {
+    /// Create a new TimeoutService that will timeout after the duration specified in `params` elapses
+    pub fn new(inner: InnerService, params: Option<TimeoutServiceParams>) -> Self {
+        Self { inner, params }
+    }
+
+    /// Create a new TimeoutService that will never timeout
+    pub fn no_timeout(inner: InnerService) -> Self {
+        Self {
+            inner,
+            params: None,
+        }
+    }
+}
```

Comment:
I can definitely go either way, but when there is only a single generic, we may want to name it `S` to follow conventions for services vended by tower directly